lemur/plugins/lemur_gcp/plugin.py

```diff
@@ -0,0 +1,105 @@
+            "helpMessage": "Authentication method to use",
+        },
+        {
+            "name": "vaultMountPoint",
+            "type": "str",
+            "required": False,
+            "helpMessage": "Path to vault secret",
+        },
+        {
+            "name": "serviceAccountTokenPath",
+            "type": "str",
+            "required": False,
+            "helpMessage": "Path to vault secret",
+        }
+    ]
+
+    def upload(self, name, body, private_key, cert_chain, options, **kwargs):
+        """
+        Args:
+        certificate_file: path to the file with the certificate you want to create in your project.
+        private_key_file: path to the private key you used to sign the certificate with.
+        certificate_name: name for the certificate once it's created in your project.
+
+        """
+
+        try:
+            ssl_certificate_body = {
+                "name": name,
+                "certificate": body,
+                "description": "",
+                "private_key": private_key,
+            }
+            credentials = self._get_gcp_credentials(options)
+            return self._insert_gcp_certificate(
+                self.get_option("projectID", options),
+                ssl_certificate_body,
+                credentials,
+            )
+
+        except Exception as e:
+            current_app.logger.error(
+                f"Issue with uploading {name} to GCP. Action failed with the following log: {e}",
+                exc_info=True,
+            )
+            raise Exception(f"Issue uploading certificate to GCP: {e}")
+
+    def _insert_gcp_certificate(self, project_id, ssl_certificate_body, credentials):
+        return ssl_certificates.SslCertificatesClient(credentials=credentials).insert(
+            project=project_id, ssl_certificate_resource=ssl_certificate_body
+        )
+
+    def _get_gcp_credentials(self, options):
+        if self.get_option('authenticationMethod', options) == "vault":
+            # make a request to vault for GCP token
+            return self._get_gcp_credentials_from_vault(options)
+        elif self.get_option('authenticationMethod', options) == "serviceAccountToken":
+            if self.get_option('serviceAccountTokenPath', options) is not None:
+                return service_account.Credentials.from_service_account_file(
+                    self.get_option('serviceAccountTokenPath', options)
+                )
+
+        raise Exception("No supported way to authenticate with GCP")
+
+    def _get_gcp_credentials_from_vault(self, options):
+        service_token = hvac.Client(os.environ['VAULT_ADDR']) \
+            .secrets.gcp \
+            .generate_oauth2_access_token(
+            roleset="",
+            mount_point=f"{self.get_option('vaultMountPoint', options)}"
+        )["data"]["token"].rstrip(".")
+        credentials, _ = google.auth.default()  # Fetch credentials from Vault GCP secrets engine
```

Comment:
More accurately, this line is fetching the default credentials for the environment as-per https://google-auth.readthedocs.io/en/master/reference/google.auth.html. The call to `generate_oauth2_access_token` in the lines above fetch the OAuth2 token from Vault.

Ultimately how much we want to tweak these comments is up to you but as written it could be misleading for someone not familiar with GCP auth attempting to make edits to this plugin in the future.

Reply:
👍 updated this comment to closer reflect what is written in the google auth documentation.